motionplan/tpSpaceRRT.go

```diff
@@ -172,185 +205,326 @@ func (mp *tpSpaceRRTMotionPlanner) planRunner(
+			if reachedDelta <= mp.algOpts.poseSolveDist {
+				// If we've reached the goal, extract the path from the RRT trees and return
+				path := extractPath(rrt.maps.startMap, rrt.maps.goalMap, &nodePair{a: seedReached.node, b: goalReached.node}, false)
 				rrt.solutionChan <- &rrtPlanReturn{steps: path, maps: rrt.maps}
 				return
 			}
 		}
+		if iter%mp.algOpts.attemptSolveEvery == 0 {
+			// Attempt a solve; we exhaustively iterate through our goal tree and attempt to find any connection to the seed tree
+			paths := [][]node{}
+			for goalMapNode := range rrt.maps.goalMap {
+				seedReached := mp.attemptExtension(ctx, goalMapNode, rrt.maps.startMap, false)
+				if seedReached.error != nil {
+					rrt.solutionChan <- &rrtPlanReturn{planerr: seedReached.error, maps: rrt.maps}
+					return
+				}
+				if seedReached.node == nil {
+					continue
+				}
+				reachedDelta := mp.planOpts.DistanceFunc(&ik.Segment{StartPosition: seedReached.node.Pose(), EndPosition: goalMapNode.Pose()})
+				if reachedDelta <= mp.algOpts.poseSolveDist {
+					// If we've reached the goal, extract the path from the RRT trees and return
+					path := extractPath(rrt.maps.startMap, rrt.maps.goalMap, &nodePair{a: seedReached.node, b: goalMapNode}, false)
+					paths = append(paths, path)
+				}
+			}
+			if len(paths) > 0 {
+				var bestPath []node
+				bestCost := math.Inf(1)
+				for _, goodPath := range paths {
+					currCost := sumCosts(goodPath)
+					if currCost < bestCost {
+						bestCost = currCost
+						bestPath = goodPath
+					}
+				}
+				correctedPath, err := mp.rectifyPath(bestPath)
+				if err != nil {
+					rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+					return
+				}
+				rrt.solutionChan <- &rrtPlanReturn{steps: correctedPath, maps: rrt.maps}
+				return
+			}
+		}
+
+		// Get random cartesian configuration
+		randPosNode, err = mp.sample(midptNode, iter+1)
+		if err != nil {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+			return
+		}
 	}
 	rrt.solutionChan <- &rrtPlanReturn{maps: rrt.maps, planerr: errors.New("tpspace RRT unable to create valid path")}
 }
 
-// getExtensionCandidate will return either nil, or the best node on a PTG to reach the desired random node and its RRT tree parent.
+// getExtensionCandidate will return either nil, or the best node on a valid PTG to reach the desired random node and its RRT tree parent.
 func (mp *tpSpaceRRTMotionPlanner) getExtensionCandidate(
 	ctx context.Context,
 	randPosNode node,
 	ptgNum int,
 	curPtg tpspace.PTG,
-	rrt *rrtParallelPlannerShared,
+	rrt rrtMap,
 	nearest node,
-	planOpts *plannerOptions, // Need to pass this in explicitly for smoothing
-) *candidate {
-	nm := &neighborManager{nCPU: mp.planOpts.NumThreads}
+	invert bool,
+) (*candidate, error) {
+	nm := &neighborManager{nCPU: mp.planOpts.NumThreads / len(mp.tpFrame.PTGs())}
```

Comment:
[nit] i don't think this is what we want to be doing
suppose `mp.planOpts.NumThreads = 1` and `len(mp.tpFrame.PTGs()) = 10`
then `1/10=0`

Reply:
This is okay to do. The behavior of NN is that it will run in parallel if nCPU is > 1; if 1, 0, or negative, it will simply run in single-threaded mode.